### PR TITLE
Storage pmd11

### PR DESCRIPTION
**Pull request by jtabarez**
The storage constraints are added to the gfmi constraints and the storage constraints will only focus on energy.

### CODE REVIEW

**Review by pseudocubic — CHANGES_REQUESTED** [review]
For docstrings, I suggest the following format to match PowerModelsDistribution's docstring format conventions:

```julia
@doc raw"""
    <function_name>(<args>; <kwargs>...)::<return_type>

<Description of function>

<math_if_applicable>
"""
```

That is for exported functions. For private functions (those starting with `_`), just a single line description will suffice. the `@doc raw"""` part is only needed if you are using LaTeX math, because of escape characters

File: src/core/constraint_inverter.jl [hunk 1]
```diff
@@ -440,3 +440,19 @@ function constraint_mc_i_inverter(pm::_PMD.AbstractUnbalancedIVRModel, n::Int, i
...
+    vi =  _PMD.var(pm, nw, :vi, bus_id)
+
+    crs =   _PMD.var(pm, nw, :crs, i)
+    cis =   _PMD.var(pm, nw, :cis, i)
+ 
+    p =  _PMD.var(pm, nw, :p_storage, i)
+    q =  _PMD.var(pm, nw, :q_storage, i)
+
+    # DC-link power
+    JuMP.@NLconstraint(pm.model, sum(vr[c]*crs[c] + vi[c]*cis[c] for c in connections) == p)
```
[the remark this answers is not in the document]
Comment (bluejuniper) [reply src/core/constraint_inverter.jl:457]: This is actually a bilinear constraint. I assume that `@NLconstraint` is needed in this case?

File: Project.toml [hunk 2]
```diff
@@ -6,6 +6,7 @@ version = "0.3.0"
 [deps]
 InfrastructureModels = "2030c09a-7f63-5d83-885d-db604e0e9cc0"
 JuMP = "4076af6c-e467-56ae-b986-b466b2749572"
+Juniper = "2ddba703-00a4-53a7-87a5-e8b9971dde84"
```
[the remark this answers is not in the document]
Comment (bluejuniper) [reply Project.toml:9]: Checked and we are not using Juniper - it's removed from deps & extras

File: src/core/constraint_inverter.jl [hunk 3]
```diff
@@ -440,3 +440,19 @@ function constraint_mc_i_inverter(pm::_PMD.AbstractUnbalancedIVRModel, n::Int, i
         JuMP.@NLconstraint(pm.model, cmax^2 == crg[c]^2 + cig[c]^2)
     end
 end
+
+
+function constraint_mc_storage_grid_forming_inverter(pm::_PMD.AbstractUnbalancedIVRModel, nw::Int, i, bus_id, vr0, vi0, pmax, qmax, qmin, cmax, smax, energy, energy_rating, ang, connections)
```
[the remark this answers is not in the document]
Comment (bluejuniper) [reply src/core/constraint_inverter.jl:445]: The unused arguments were left over from before the inverter-only constraints from `constraint_mc_storage_grid_forming_inverter` were split out into `constraint_mc_grid_forming_inverter`. Removed

File: src/core/variable.jl [hunk 4]
```diff
@@ -324,3 +324,81 @@ function variable_mc_bus_fault_current(pm::_PMD.AbstractUnbalancedIVRModel; nw::
...
+            start = _PMD.comp_start_value(_PMD.ref(pm, nw, :storage, i), "cis_start", c, 0.0)
+        ) for i in _PMD.ids(pm, nw, :storage)
+    )
+    if bounded
+    end
+end
+
+
+function variable_mc_storage_grid_forming_inverter(pm::_PMD.AbstractUnbalancedIVRModel; nw::Int=nw_id_default, bounded::Bool=true, kwargs...)
+    connections = Dict(i => storage["connections"] for (i,storage) in _PMD.ref(pm, nw, :storage))
```
[the remark this answers is not in the document]
Comment (bluejuniper) [reply src/core/variable.jl:358]: If I'm following correctly the `bus_conns_storage` dictionary provides a list of tuples for each BESS at a bus so it's not exactly the data structure we are using

File: src/core/constraint_inverter.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (pseudocubic) [reply src/core/constraint_inverter.jl:457]: I'm pretty sure quadratic and bilinear constraints can still use `@constraint`, and don't require `@NLconstraint`

File: src/core/constraint_template.jl [hunk 5]
```diff
@@ -426,3 +426,27 @@ function constraint_mc_bus_fault_current(pm::_PMD.AbstractUnbalancedIVRModel, i:
 
     constraint_mc_bus_fault_current(pm, nw, i, bus, connections, Gf, Bf)
 end
+
+
+""
+function constraint_mc_storage_grid_forming_inverter(pm::_PMD.AbstractUnbalancedIVRModel, i::Int; nw::Int=nw_id_default)
+    storage = _PMD.ref(pm, nw, :storage, i)
+    baseMVA = _PMD.ref(pm, nw, :settings, "sbase")
+    haskey(storage, "i_max") ? nothing : storage["i_max"] = 1.1*storage["dss"]["kva"] / (3 * 100 * baseMVA)
```
[the remark this answers is not in the document]
Comment (bluejuniper) [reply src/core/constraint_template.jl:435]: Removed this as it's no longer being used anyways

**Review by bluejuniper — APPROVED** [review]
Finishing going through David's comments and added some more unit tests. Tests are passing

File: Project.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (pseudocubic) [reply Project.toml:9]: It seems to still be appearing in the latest changes

**Comment by pseudocubic** [comment]
There is still one issue from the code comments (Juniper still appears in deps in the project.toml), I think you're missing an updated changelog, and I still recommend going through and updating the docstrings to the suggested format, at least for the new functions you added. 

other than that all the other issues seem resolved.

**Review by pseudocubic — APPROVED** [review]

**Pull request merged**